File: bapc2015/mario.py
for _ in range(int(input())):
    n, w = map(int, input().split())
    
    l = [0] * n
    r = [0] * n
    for i in range(n):
        l[i], r[i] = map(int, input().split())
    
    x = 0
    direction = [1] * n
    pos = [0] * n
    for i in range(n):
        pos[i] = l[i]
    for it in range(2 * w * w):
        fwd = [False] * (2 * w + 1)
        for i in range(n):
            if direction[i] == 1:
                fwd[pos[i]] = True
            pos[i] += direction[i]
            if pos[i] == l[i] or pos[i] == r[i]:
                direction[i] *= -1
        
        if fwd[x]:
            x += 1
        elif not fwd[x - 1]:
            x -= 1
        
        if x == w:
            print(it + 1)
            break
    else:
        print("IMPOSSIBLE")

File: bapc2015/mario.cpp
#include <bits/stdc++.h>

#define int long long

using namespace std;

signed main() {
    int t; cin >> t;
    while (t--) {
        int n, w; cin >> n >> w;
        int l[n], r[n];
        int pos[n];
        int direction[n];
        for (int i = 0; i < n; i++) {
            cin >> l[i] >> r[i];
            pos[i] = l[i];
            direction[i] = 1;
        }
        int x = 0;
        int time = 0;
        while (x < w and time++ < w * w * 2) {
            vector<bool> forward(w * 4, false);

            for (int i = 0; i < n; i++) {
                if (direction[i] == 1)
                    forward[pos[i]] = true;
                pos[i] += direction[i];
                if (pos[i] == l[i] || pos[i] == r[i])
                    direction[i] *= -1;
            }
            if (forward[x]) x++;
            else if (!forward[x - 1]) x--;
        }
        if (x == w)
            cout << time << endl;
        else
            cout << "IMPOSSIBLE" << endl;



    }
}